### PR TITLE
Migrate to Swift 3 Beta 6

### PR DESCRIPTION
**Pull request by JadenGeller**
This fixes the Swift compilation issues in Realm Swift and Realm Objective-C. Note this PR makes the smallest possible changes to get the current codebase to compile. After this is merged, we should also improve the API by considering the following:
- Support Swift native `Date` and `Data` types (https://github.com/realm/realm-cocoa/issues/3968)
- Improve type-safety of Swift 3 API (https://github.com/realm/realm-cocoa/issues/3969)
- Find better way to specify `NSNumber` type from Swift when using Realm Objective-C (https://github.com/realm/realm-cocoa/issues/3971)
- Use `Optional`s instead of `NSNull` in Swift API (https://github.com/realm/realm-cocoa/issues/3972)
- Create protocol for types that can be stored as `Realm` object properties (https://github.com/realm/realm-cocoa/issues/3973)
- Vet Realm Swift API `open` members (https://github.com/realm/realm-cocoa/issues/3974)
- Fix segfaulting Swift functions in Realm Objective-C (https://github.com/realm/realm-cocoa/issues/3975)
- Introduce `addObjects` Swift helper function in Realm Objective-C (https://github.com/realm/realm-cocoa/issues/3976)

### CODE REVIEW

**Comment by JadenGeller** [comment]
Note this PR adds some annotations to the Objective-C API to improve import into Swift. If this is deemed undesirable, we can forgo those changes, but I can't imagine them breaking any reasonable code.

**Comment by raphaelcruzeiro** [comment]
Hey @JadenGeller is there anyway I can build realm from this PR or is it better if I wait for it to be merged? I get a lot C++ issues when I try to build from this branch (maybe I am missing some steps during my build process?). 

Anyway, sorry for generating noise here.

**Comment by JadenGeller** [comment]
@raphaelcruzeiro You should be able to build from this branch. Note that we've moved the [Object Store](https://github.com/realm/realm-object-store) to a submodule, so you'll need to `git submodule update` after checking out the branch.

**Comment by JadenGeller** [comment]
@jpsim @austinzheng @tgoyne @bdash This is ready for review! 🎉

File: Realm/Swift/RLMSupport.swift [hunk 1]
```diff
@@ -20,14 +20,15 @@ import Realm
...
     extension RLMRealm {
-        public class func schemaVersion(at url: URL, usingEncryptionKey key: Data? = nil) throws -> UInt64 {
-            var error: NSError?
-            let version = __schemaVersion(at: url, encryptionKey: key, error: &error)
-            guard version != RLMNotVersioned else {
-                throw error!
-            }
-            return version
-        }
+        // TODO: Figure out why this causes the Swift 3 compiler to segfault.
```
Comment (bdash) [inline Realm/Swift/RLMSupport.swift]: Have you filed a Swift bug about this crash?

File: RealmSwift/List.swift [hunk 2]
```diff
@@ -491,8 +491,8 @@ extension List : RealmCollection, RangeReplaceableCollection {
     - parameter subRange:    The range of elements to be replaced.
     - parameter newElements: The new elements to be inserted into the List.
     */
-    public func replaceSubrange<C : Collection where C.Iterator.Element == T>(_ subrange: Range<Int>,
-                                                                              with newElements: C) {
+    public func replaceSubrange<C : Collection>(_ subrange: Range<Int>,
+                                with newElements: C) where C.Iterator.Element == T {
```
Comment (bdash) [inline RealmSwift/List.swift]: The indentation of this line looks wonky.

File: RealmSwift/Migration.swift [hunk 3]
```diff
@@ -192,7 +193,7 @@ internal func accessorMigrationBlock(_ migrationBlock: MigrationBlock) -> RLMMig
         }
 
         // run migration
-        migrationBlock(migration: Migration(migration), oldSchemaVersion: oldVersion)
+        migrationBlock(/* migration: */ Migration(migration), /* oldSchemaVersion: */ oldVersion)
```
Comment (bdash) [inline RealmSwift/Migration.swift]: I don't think these comments add value.

File: RealmSwift/Migration.swift [hunk 4]
```diff
@@ -203,7 +204,7 @@ extension Migration {
     public func enumerate(_ objectClassName: String, _ block: MigrationObjectEnumerateBlock) { }
 
     @available(*, unavailable, renamed:"createObject(ofType:populatedWith:)")
-    public func create(_ className: String, value: AnyObject = [:]) -> MigrationObject {
+    public func create(_ className: String, value: AnyObject = [:] as NSDictionary) -> MigrationObject {
```
Comment (bdash) [inline RealmSwift/Migration.swift]: Is `AnyObject` still correct here?

File: Realm/Swift/RLMSupport.swift (hunk 1 above)
Comment (JadenGeller) [inline Realm/Swift/RLMSupport.swift]: Not yet. Still don't know what causes it.

File: RealmSwift/Optional.swift [hunk 5]
```diff
@@ -31,34 +31,33 @@ extension Float: RealmOptionalType {}
 extension Double: RealmOptionalType {}
 extension Bool: RealmOptionalType {}
 
-// Not all RealmOptionalType's can be cast to AnyObject, so handle casting logic here.
-private func realmOptionalToAnyObject<T: RealmOptionalType>(value: T?) -> AnyObject? {
-    if let anyObjectValue: AnyObject = value as? AnyObject {
-        return anyObjectValue
-    } else if let int8Value = value as? Int8 {
+private func realmOptionalToAny<T: RealmOptionalType>(_ value: T) -> Any {
+    // FIXME: Use common protocol that defines bridging instead of special-case check with no exaustiveness guarentees.
```
Comment (bdash) [inline RealmSwift/Optional.swift]: exhaustiveness

File: RealmSwift/Optional.swift [hunk 6]
```diff
@@ -31,34 +31,33 @@ extension Float: RealmOptionalType {}
...
+    } else {
+        return value
     }
-    return nil
 }
 
-// Not all RealmOptionalType's can be cast from AnyObject, so handle casting logic here.
-private func anyObjectToRealmOptional<T: RealmOptionalType>(anyObject: AnyObject?) -> T? {
+private func anyToRealmOptional<T: RealmOptionalType>(_ value: Any) -> T {
+    // FIXME: Use common protocol that defines bridging instead of special-case check with no exaustiveness guarentees.
```
Comment (bdash) [inline RealmSwift/Optional.swift]: exhaustiveness

File: RealmSwift/Migration.swift (hunk 3 above)
Comment (JadenGeller) [inline RealmSwift/Migration.swift]: Ok. I'm hoping that eventually lambda variables can be named `migrationBlock(migration:oldSchemaVersion:)` so we get our argument labels back.

File: RealmSwift/RealmCollectionType.swift [hunk 7]
```diff
@@ -756,8 +756,8 @@ public final class AnyRealmCollection<T: Object>: RealmCollection {
 
     - returns: `Results` with elements sorted by the given sort descriptors.
     */
-    public func sorted<S: Sequence where S.Iterator.Element == SortDescriptor>
-                      (with sortDescriptors: S) -> Results<Element> {
+    public func sorted<S: Sequence>
+        (with sortDescriptors: S) -> Results<Element> where S.Iterator.Element == SortDescriptor {
```
Comment (bdash) [inline RealmSwift/RealmCollectionType.swift]: This feels like it is wrapped at a weird place. Why not after the return type?

File: RealmSwift/Migration.swift (hunk 4 above)
Comment (JadenGeller) [inline RealmSwift/Migration.swift]: I figure we ought not change the APIs that are marked unavailable, no? They're only there to assist in migration.

File: RealmSwift/Tests/MigrationTests.swift [hunk 8]
```diff
@@ -63,7 +63,7 @@ class MigrationTests: TestCase {
         let config = Realm.Configuration(fileURL: fileURL, schemaVersion: schemaVersion,
             migrationBlock: { migration, oldSchemaVersion in
                 if let block = block {
-                    block(migration: migration, oldSchemaVersion: oldSchemaVersion)
+                    block(/* migration: */ migration, /* oldSchemaVersion: */ oldSchemaVersion)
```
Comment (bdash) [inline RealmSwift/Tests/MigrationTests.swift]: Comments seem pointless.

File: RealmSwift/RealmCollectionType.swift (hunk 7 above)
Comment (JadenGeller) [inline RealmSwift/RealmCollectionType.swift]: I don't know how this happened ._.

File: RealmSwift/Migration.swift [hunk 9]
```diff
@@ -109,8 +109,8 @@ public final class Migration {
     */
     public func enumerateObjects(ofType typeName: String, _ block: MigrationObjectEnumerateBlock) {
         rlmMigration.enumerateObjects(typeName) {
-            block(oldObject: unsafeBitCast($0, to: MigrationObject.self),
-                  newObject: unsafeBitCast($1, to: MigrationObject.self))
+            block(/* oldObject: */ unsafeBitCast($0, to: MigrationObject.self),
```
Comment (JadenGeller) [inline RealmSwift/Migration.swift]: @bdash Do you find this equally unuseful?

Comment (bdash) [inline RealmSwift/Migration.swift]: This case is marginally useful because the arguments do not have clear names.

Comment (tgoyne) [inline RealmSwift/Migration.swift]: Giving the arguments names seems better than inline comments naming them.

**Comment by bdash** [comment]
👍

File: RealmSwift/Object.swift [hunk 10]
```diff
@@ -450,7 +450,7 @@ public class Object: RLMObjectBase {
      - parameter value:  The value used to populate the object.
     */
     public init(value: AnyObject) {
-        self.dynamicType.sharedSchema() // ensure this class' objectSchema is loaded in the partialSharedSchema
+    self.dynamicType.sharedSchema() // ensure this class' objectSchema is loaded in the partialSharedSchema
```
Comment (tgoyne) [inline RealmSwift/Object.swift]: incorrect indentation change

Comment (JadenGeller) [inline RealmSwift/Object.swift]: Good catch. How does this happen!? 👀

**Comment by tgoyne** [comment]
👍

**Pull request merged**